amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShutdownTask.java

```diff
@@ -96,68 +103,57 @@ public TaskResult call() {
         Exception exception;
         boolean applicationException = false;
 
+        final ShutdownReason localReason;
+        // If shut down reason is SHARD_END but childShards is empty list, this means we get to a false shard end.
+        // We should not throw exception because it will trigger another ShutdownTask which has the same empty list of childShards.
+        // Instead we should shut down the ShardConsumer with LEASE_LOST reason to allow other worker to grab lease and continue reading from this shard.
+        if(reason == ShutdownReason.SHARD_END && CollectionUtils.isNullOrEmpty(childShards)) {
+            localReason = ShutdownReason.LEASE_LOST;
+            dropLease();
+            log.warn("False shard end encountered. Forcing the lease to be lost before shutting down the consumer for Shard: {}.", shardInfo.shardId());
+        } else {
+            localReason = reason;
+        }
+
         try {
             try {
-                ShutdownReason localReason = reason;
-                List<Shard> latestShards = null;
-                /*
-                 * Revalidate if the current shard is closed before shutting down the shard consumer with reason SHARD_END
-                 * If current shard is not closed, shut down the shard consumer with reason LEASE_LOST that allows
-                 * active workers to contend for the lease of this shard.
-                 */
-                if (localReason == ShutdownReason.SHARD_END) {
-                    latestShards = shardDetector.listShards();
-
-                    //If latestShards is empty, should also shutdown the ShardConsumer without checkpoint with SHARD_END
-                    if (CollectionUtils.isNullOrEmpty(latestShards) || !isShardInContextParentOfAny(latestShards)) {
-                        localReason = ShutdownReason.LEASE_LOST;
-                        dropLease();
-                        log.info("Forcing the lease to be lost before shutting down the consumer for Shard: " + shardInfo.shardId());
-                    }
-                }
+                log.debug("Invoking shutdown() for shard {}, concurrencyToken {}. Shutdown reason: {}",
+                        shardInfo.shardId(), shardInfo.concurrencyToken(), localReason);
 
-                // If we reached end of the shard, set sequence number to SHARD_END.
+                final long startTime = System.currentTimeMillis();
                 if (localReason == ShutdownReason.SHARD_END) {
+                    // Create new lease for the child shards if they don't exist.
+                    createLeasesForChildShardsIfNotExist();
```

Comment:
Why is this not part of inner try. We should drop the lease if we are unable to create leases for the child shards. So that the next worker will resume from previous checkpoint.

Reply:
As discussed offline, this call is not in the inner try because it's not the customer implemented code (recordProcessorChechpointer). The inner try should only catch exceptions from customer implemented code. And we also decided not to drop the lease the lease, instead, we will let the same worker handle the shutdown in the next ShutdownTask. The childShards information is cached in the ShardConsumer so that the next ShutdownTask would be created without losing the childShards info.